Can not run pytest directly on notebook as suggested in doc

### Describe the bug

```
pytest marimo_test.py
```

works as expected with marimo 0.12.0 

```
pytest marimo_test.py  
================================================== test session starts ===================================================
platform linux -- Python 3.12.9, pytest-8.3.4, pluggy-1.5.0 -- /home/remi-gau/miniconda3/bin/python3
cachedir: .pytest_cache
Using --randomly-seed=812283614
rootdir: /home/remi-gau/github/poia
configfile: pyproject.toml
plugins: reporter-0.5.3, cov-6.0.0, timeout-2.3.1, anyio-4.9.0, xdist-3.6.1, reporter-html1-0.9.2, nbmake-1.5.5, randomly-3.16.0, csv-3.0.0
collected 4 items                                                                                                        

marimo_test.py::test_sanity PASSED
marimo_test.py::test_fails FAILED
marimo_test.py::MarimoTestBlock_0::test_answer[4-5] PASSED
marimo_test.py::MarimoTestBlock_0::test_answer[3-4] PASSED

======================================================== FAILURES ========================================================
_______________________________________________________ test_fails _______________________________________________________

    # /// script
    # requires-python = ">=3.12"
    # dependencies = [
    #     "marimo",
    #     "pytest==8.3.5",
    # ]
    # ///
    import marimo
    
    __generated_with = "0.12.5"
    app = marimo.App()
    
    
    @app.cell
    def _():
        def inc(x):
            return x + 1
        return (inc,)
    
    
    @app.cell
    def test_fails(inc):
>       assert inc(3) == 5, "This test fails"
E       AssertionError: This test fails
E       assert 4 == 5
E        +  where 4 = <function inc at 0x759bf94d9c60>(3)

marimo_test.py:23: AssertionError
================================================ short test summary info =================================================
FAILED marimo_test.py::test_fails - AssertionError: This test fails
assert 4 == 5
 +  where 4 = <function inc at 0x759bf94d9c60>(3)
============================================== 1 failed, 3 passed in 0.69s ===============================================
```

but with marimo > 0.12.0

```
pytest marimo_test.py     
================================================== test session starts ===================================================
platform linux -- Python 3.12.9, pytest-8.3.4, pluggy-1.5.0 -- /home/remi-gau/miniconda3/bin/python3
cachedir: .pytest_cache
Using --randomly-seed=2704945492
rootdir: /home/remi-gau/github/poia
configfile: pyproject.toml
plugins: reporter-0.5.3, cov-6.0.0, timeout-2.3.1, anyio-4.9.0, xdist-3.6.1, reporter-html1-0.9.2, nbmake-1.5.5, randomly-3.16.0, csv-3.0.0
collected 4 items                                                                                                        

marimo_test.py::MarimoTestBlock_0::test_answer[4-5] PASSED
marimo_test.py::MarimoTestBlock_0::test_answer[3-4] PASSED
marimo_test.py::test_fails FAILED
marimo_test.py::test_sanity FAILED

======================================================== FAILURES ========================================================
_______________________________________________________ test_fails _______________________________________________________

self = Cell(_name='test_fails', _cell=CellImpl(key=-6658325269729851256, code='assert inc(3) == 5, "This test fails"', mod=<a....app.InternalApp object at 0x7c04e5b3ddc0>, _pytest_reserved={'inc'}, _test_allowed=True, _expected_signature=('inc',))
args = (), kwargs = {}, TopLevelExtraction = <class 'marimo._ast.toplevel.TopLevelExtraction'>
extraction = <marimo._ast.toplevel.TopLevelExtraction object at 0x7c04a9152360>, arg_names = [], argc = 0, call_argc = 1
unexpected = set(), is_pytest = True

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        from marimo._ast.toplevel import TopLevelExtraction
    
        assert self._app is not None
        # Definitions on a module level are not part of the signature, and as
        # such, should not be provided with the call.
        extraction = TopLevelExtraction.from_app(self._app)
        arg_names = sorted(
            (self._cell.refs - extraction.allowed_refs) - self._cell.defs
        )
        argc = len(arg_names)
    
        call_args = {name: arg for name, arg in zip(arg_names, args)}
        call_args.update(kwargs)
        call_argc = len(call_args.keys()) + len(self._pytest_reserved)
    
        if unexpected := call_args.keys() - arg_names:
            raise TypeError(
                f"{self.name}() got an unexpected argument(s) '{unexpected}'"
            )
    
        is_pytest = "PYTEST_CURRENT_TEST" in os.environ
        # Capture pytest case, where arguments don't match the references.
        if self._pytest_reserved - set(arg_names):
>           raise TypeError(
                "A mismatch in expected argument names likely means you should "
                "resave the notebook in the marimo editor."
            )
E           TypeError: A mismatch in expected argument names likely means you should resave the notebook in the marimo editor.

../../miniconda3/lib/python3.12/site-packages/marimo/_ast/cell.py:635: TypeError
______________________________________________________ test_sanity _______________________________________________________

self = Cell(_name='test_sanity', _cell=CellImpl(key=-7109420379092606081, code='assert inc(3) == 4, "This test passes"', mod=....app.InternalApp object at 0x7c04a90ad4c0>, _pytest_reserved={'inc'}, _test_allowed=True, _expected_signature=('inc',))
args = (), kwargs = {}, TopLevelExtraction = <class 'marimo._ast.toplevel.TopLevelExtraction'>
extraction = <marimo._ast.toplevel.TopLevelExtraction object at 0x7c04a910ec60>, arg_names = [], argc = 0, call_argc = 1
unexpected = set(), is_pytest = True

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        from marimo._ast.toplevel import TopLevelExtraction
    
        assert self._app is not None
        # Definitions on a module level are not part of the signature, and as
        # such, should not be provided with the call.
        extraction = TopLevelExtraction.from_app(self._app)
        arg_names = sorted(
            (self._cell.refs - extraction.allowed_refs) - self._cell.defs
        )
        argc = len(arg_names)
    
        call_args = {name: arg for name, arg in zip(arg_names, args)}
        call_args.update(kwargs)
        call_argc = len(call_args.keys()) + len(self._pytest_reserved)
    
        if unexpected := call_args.keys() - arg_names:
            raise TypeError(
                f"{self.name}() got an unexpected argument(s) '{unexpected}'"
            )
    
        is_pytest = "PYTEST_CURRENT_TEST" in os.environ
        # Capture pytest case, where arguments don't match the references.
        if self._pytest_reserved - set(arg_names):
>           raise TypeError(
                "A mismatch in expected argument names likely means you should "
                "resave the notebook in the marimo editor."
            )
E           TypeError: A mismatch in expected argument names likely means you should resave the notebook in the marimo editor.

../../miniconda3/lib/python3.12/site-packages/marimo/_ast/cell.py:635: TypeError
================================================ short test summary info =================================================
FAILED marimo_test.py::test_fails - TypeError: A mismatch in expected argument names likely means you should resave the notebook in the marimo editor.
FAILED marimo_test.py::test_sanity - TypeError: A mismatch in expected argument names likely means you should resave the notebook in the marimo editor.
============================================== 2 failed, 2 passed in 0.73s ===============================================
```

### Environment

You are not going to like this, but I don't the error is due to me not having chrome.

<details>

```
Traceback (most recent call last):
  File "/home/remi-gau/miniconda3/bin/marimo", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/click/core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/click/core.py", line 1082, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/click/core.py", line 1697, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/click/core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/click/core.py", line 788, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/marimo/_cli/cli.py", line 860, in env
    click.echo(json.dumps(get_system_info(), indent=2))
                          ^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/marimo/_cli/envinfo.py", line 61, in get_system_info
    "Browser": get_chrome_version() or "--",
               ^^^^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/marimo/_utils/health.py", line 142, in get_chrome_version
    return get_chrome_version_linux()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/site-packages/marimo/_utils/health.py", line 122, in get_chrome_version_linux
    process = subprocess.Popen(
              ^^^^^^^^^^^^^^^^^
  File "/home/remi-gau/miniconda3/lib/python3.12/subprocess.py", line 1028, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/remi-gau/miniconda3/lib/python3.12/subprocess.py", line 1963, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
NotADirectoryError: [Errno 20] Not a directory: 'google-chrome'
```

</details>


### Code to reproduce

Adapted from https://docs.marimo.io/guides/testing/pytest/#example

Had to tweak the following to parametrized variables to match the test function paramaters.

```python
@app.cell
def collection_of_tests(inc, pytest):
    @pytest.mark.parametrize("input, expected", [(3, 4), (4, 5)])
    def test_answer(x, y):
        assert inc(x) == y, "These tests should pass."
```

```
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "marimo",
#     "pytest==8.3.5",
# ]
# ///
import marimo

__generated_with = "0.12.5"
app = marimo.App()


@app.cell
def _():
    def inc(x):
        return x + 1
    return (inc,)


@app.cell
def test_fails(inc):
    assert inc(3) == 5, "This test fails"


@app.cell
def test_sanity(inc):
    assert inc(3) == 4, "This test passes"

@app.cell
def collection_of_tests(inc, pytest):
    @pytest.mark.parametrize("input, expected", [(3, 4), (4, 5)])
    def test_answer(input, expected):
        assert inc(input) == expected, "These tests should pass."

@app.cell
def imports():
    import pytest
    return pytest
```

--- a/docs/guides/testing/pytest.md
+++ b/docs/guides/testing/pytest.md
@@ -57,7 +57,7 @@ def test_sanity(inc):
 
 @app.cell
 def collection_of_tests(inc, pytest):
-    @pytest.mark.parametrize("input, expected", [(3, 4), (4, 5)])
+    @pytest.mark.parametrize(("x", "y"), [(3, 4), (4, 5)])
     def test_answer(x, y):
         assert inc(x) == y, "These tests should pass."
 
@@ -71,7 +71,7 @@ prints
 
 ```pytest
 ============================= test session starts ==============================
-platform linux -- Python 3.11.10, pytest-8.3.4, pluggy-1.5.0
+platform linux -- Python 3.12.9, pytest-8.3.5, pluggy-1.5.0
 rootdir: /notebooks
 configfile: pyproject.toml
 collected 4 items
@@ -108,7 +108,5 @@ E        +  where 4 = <function inc>(3)
 test_notebook.py:17: AssertionError
 =========================== short test summary info ============================
 FAILED test_notebook.py::test_fails - AssertionError: This test fails
-assert 4 == 5
- +  where 4 = <function inc>(3)
-========================= 1 failed, 3 passed in 0.65s ==========================
+========================= 1 failed, 3 passed in 0.82s ==========================
 ```

--- a/marimo/_ast/cell.py
+++ b/marimo/_ast/cell.py
@@ -620,10 +620,13 @@ class Cell:
         assert self._app is not None
         # Definitions on a module level are not part of the signature, and as
         # such, should not be provided with the call.
-        extraction = TopLevelExtraction.from_app(self._app)
-        arg_names = sorted(
-            (self._cell.refs - extraction.allowed_refs) - self._cell.defs
-        )
+
+        # NB. TopLevelExtraction assumes that all cells that can be exposed will
+        # be, but signature provides context for what is actually scoped.
+        allowed_refs = TopLevelExtraction.from_app(self._app).allowed_refs
+        allowed_refs -= set(self._expected_signature or ())
+
+        arg_names = sorted((self._cell.refs - allowed_refs) - self._cell.defs)
         argc = len(arg_names)
 
         call_args = {name: arg for name, arg in zip(arg_names, args)}

--- a/tests/_ast/test_pytest_scoped.py
+++ b/tests/_ast/test_pytest_scoped.py
@@ -19,5 +19,20 @@ def _(test_cases):
     return (test_function,)
 
 
+@app.cell
+def _():
+    def inc(x):
+        return x + 1
+
+    return (inc,)
+
+
+@app.cell
+def collection_of_tests(inc, pytest):
+    @pytest.mark.parametrize(("x", "y"), [(3, 4), (4, 5)])
+    def test_answer(x, y):
+        assert inc(x) == y, "These tests should pass."
+
+
 if __name__ == "__main__":
     app.run()

